File: AyantDomain/Sources/AyantDomain/Tetris.swift
import Foundation

/// Правила тетриса — чистые и общие для обеих платформ.
///
/// Игра живёт в домене по той же причине, что `Ranking` и `PointsMath`: это
/// арифметика без экрана, её надо считать одинаково на iOS и Android и покрывать
/// тестами. Вью остаётся только нарисовать `State` и слать ходы.
///
/// Прошлая версия тетриса была на SpriteKit, у которого нет пары в Compose —
/// поэтому её и убрали. Здесь только сетка, так что обе платформы рисуют её
/// штатными средствами.
public enum Tetris {
    public static let columns = 10
    public static let rows = 20

    /// Сколько бонусов даёт одна линия. Начисление всё равно проходит через
    /// общий дневной лимит мини-игр — своей экономики у игры нет.
    public static let bonusPerLine = 5

    public enum Shape: Int, CaseIterable, Sendable {
        case i, o, t, s, z, j, l
    }

    /// Занятые клетки фигуры в её рамке 4×4 для каждого поворота.
    public static func cells(_ shape: Shape, rotation: Int) -> [(x: Int, y: Int)] {
        let r = ((rotation % 4) + 4) % 4
        switch shape {
        case .i:  return r % 2 == 0 ? [(0,1),(1,1),(2,1),(3,1)] : [(2,0),(2,1),(2,2),(2,3)]
        case .o:  return [(1,0),(2,0),(1,1),(2,1)]
        case .s:  return r % 2 == 0 ? [(1,0),(2,0),(0,1),(1,1)] : [(1,0),(1,1),(2,1),(2,2)]
        case .z:  return r % 2 == 0 ? [(0,0),(1,0),(1,1),(2,1)] : [(2,0),(1,1),(2,1),(1,2)]
        case .t:
            switch r {
            case 0: return [(1,0),(0,1),(1,1),(2,1)]
            case 1: return [(1,0),(1,1),(2,1),(1,2)]
            case 2: return [(0,1),(1,1),(2,1),(1,2)]
            default: return [(1,0),(0,1),(1,1),(1,2)]
            }
        case .j:
            switch r {
            case 0: return [(0,0),(0,1),(1,1),(2,1)]
            case 1: return [(1,0),(2,0),(1,1),(1,2)]
            case 2: return [(0,1),(1,1),(2,1),(2,2)]
            default: return [(1,0),(1,1),(0,2),(1,2)]
            }
        case .l:
            switch r {
            case 0: return [(2,0),(0,1),(1,1),(2,1)]
            case 1: return [(1,0),(1,1),(1,2),(2,2)]
            case 2: return [(0,1),(1,1),(2,1),(0,2)]
            default: return [(0,0),(1,0),(1,1),(1,2)]
            }
        }
    }

    public struct Piece: Equatable, Sendable {
        public var shape: Shape
        public var rotation: Int
        /// Левый верхний угол рамки 4×4 в координатах поля.
        public var x: Int
        public var y: Int

        public init(shape: Shape, rotation: Int = 0, x: Int = 3, y: Int = -1) {
            self.shape = shape; self.rotation = rotation; self.x = x; self.y = y
        }

        /// Занятые клетки в координатах поля.
        public var occupied: [(x: Int, y: Int)] {
            Tetris.cells(shape, rotation: rotation).map { (x + $0.x, y + $0.y) }
        }
    }

    public struct State: Equatable, Sendable {
        /// Уложенные клетки: `rows` × `columns`, `nil` — пусто.
        public var board: [[Shape?]]
        public var piece: Piece?
        public var next: Shape
        public var lines: Int
        public var isOver: Bool
        /// Свой генератор, чтобы состояние оставалось воспроизводимым в тестах.
        public var seed: UInt64

        public init(board: [[Shape?]], piece: Piece?, next: Shape,
                    lines: Int, isOver: Bool, seed: UInt64) {
            self.board = board; self.piece = piece; self.next = next
            self.lines = lines; self.isOver = isOver; self.seed = seed
        }

        /// Заработанные бонусы (до дневного лимита — его считает `BonusEngine`).
        public var bonuses: Int { lines * Tetris.bonusPerLine }
    }

    // MARK: - Старт и случайность

    public static func start(seed: UInt64 = 0x9E3779B97F4A7C15) -> State {
        var s = seed
        let first = randomShape(&s)
        let next = randomShape(&s)
        let empty = [[Shape?]](repeating: [Shape?](repeating: nil, count: columns), count: rows)
        return State(board: empty, piece: Piece(shape: first), next: next,
                     lines: 0, isOver: false, seed: s)
    }

    /// Линейный конгруэнтный генератор: нужен воспроизводимый, а не «настоящий»
    /// рандом — иначе тест на конкретную раскладку не написать.
    private static func randomShape(_ seed: inout UInt64) -> Shape {
        seed = seed &* 6_364_136_223_846_793_005 &+ 1_442_695_040_888_963_407
        return Shape(rawValue: Int((seed >> 33) % UInt64(Shape.allCases.count))) ?? .i
    }

    // MARK: - Проверки

    /// Помещается ли фигура: в поле по бокам/снизу и не наезжает на уложенное.
    public static func fits(_ piece: Piece, in board: [[Shape?]]) -> Bool {
        for cell in piece.occupied {
            if cell.x < 0 || cell.x >= columns || cell.y >= rows { return false }
            if cell.y < 0 { continue }          // над полем — это нормально при спавне
            if board[cell.y][cell.x] != nil { return false }
        }
        return true
    }

    // MARK: - Ходы

    public static func move(_ state: State, dx: Int) -> State {
        guard var p = state.piece, !state.isOver else { return state }
        p.x += dx
        guard fits(p, in: state.board) else { return state }
        var next = state; next.piece = p; return next
    }

    public static func rotate(_ state: State) -> State {
        guard var p = state.piece, !state.isOver else { return state }
        p.rotation = (p.rotation + 1) % 4
        // Простой «отскок» от стен: пробуем сдвинуть на клетку внутрь.
        for dx in [0, -1, 1, -2, 2] {
            var candidate = p
            candidate.x += dx
            if fits(candidate, in: state.board) {
                var next = state; next.piece = candidate; return next
            }
        }
        return state
    }

    /// Шаг вниз. Если некуда — фигура фиксируется, линии сгорают, приходит новая.
    public static func tick(_ state: State) -> State {
        guard let p = state.piece, !state.isOver else { return state }
        var dropped = p
        dropped.y += 1
        if fits(dropped, in: state.board) {
            var next = state; next.piece = dropped; return next
        }
        return lock(state, piece: p)
    }

    public static func hardDrop(_ state: State) -> State {
        guard var p = state.piece, !state.isOver else { return state }
        while true {
            var down = p
            down.y += 1
            if fits(down, in: state.board) { p = down } else { break }
        }
        return lock(state, piece: p)
    }

    private static func lock(_ state: State, piece: Piece) -> State {
        var board = state.board
        for cell in piece.occupied where cell.y >= 0 {
            board[cell.y][cell.x] = piece.shape
        }
        // Сгоревшие линии убираем, сверху досыпаем пустые.
        let kept = board.filter { row in row.contains(where: { $0 == nil }) }
        let cleared = rows - kept.count
        let empty = [Shape?](repeating: nil, count: columns)
        board = [[Shape?]](repeating: empty, count: cleared) + kept

        var seed = state.seed
        let spawned = Piece(shape: state.next)
        let following = randomShape(&seed)
        let over = !fits(spawned, in: board)

        return State(board: board,
                     piece: over ? nil : spawned,
                     next: following,
                     lines: state.lines + cleared,
                     isOver: over,
                     seed: seed)
    }
}

File: android/domain/src/main/kotlin/kg/ayant/app/domain/Tetris.kt
package kg.ayant.app.domain

/**
 * Правила тетриса — чистые и общие для обеих платформ. Зеркалит `Tetris.swift`.
 *
 * Игра живёт в домене по той же причине, что `Ranking` и `PointsMath`: это
 * арифметика без экрана, её надо считать одинаково на iOS и Android и покрывать
 * тестами. Вью остаётся только нарисовать [State] и слать ходы.
 */
object Tetris {
    const val COLUMNS = 10
    const val ROWS = 20

    /**
     * Сколько бонусов даёт одна линия. Начисление всё равно проходит через общий
     * дневной лимит мини-игр — своей экономики у игры нет.
     */
    const val BONUS_PER_LINE = 5

    enum class Shape { I, O, T, S, Z, J, L }

    data class Cell(val x: Int, val y: Int)

    /** Занятые клетки фигуры в её рамке 4×4 для каждого поворота. */
    fun cells(shape: Shape, rotation: Int): List<Cell> {
        val r = ((rotation % 4) + 4) % 4
        return when (shape) {
            Shape.I -> if (r % 2 == 0) listOf(Cell(0, 1), Cell(1, 1), Cell(2, 1), Cell(3, 1))
                       else listOf(Cell(2, 0), Cell(2, 1), Cell(2, 2), Cell(2, 3))
            Shape.O -> listOf(Cell(1, 0), Cell(2, 0), Cell(1, 1), Cell(2, 1))
            Shape.S -> if (r % 2 == 0) listOf(Cell(1, 0), Cell(2, 0), Cell(0, 1), Cell(1, 1))
                       else listOf(Cell(1, 0), Cell(1, 1), Cell(2, 1), Cell(2, 2))
            Shape.Z -> if (r % 2 == 0) listOf(Cell(0, 0), Cell(1, 0), Cell(1, 1), Cell(2, 1))
                       else listOf(Cell(2, 0), Cell(1, 1), Cell(2, 1), Cell(1, 2))
            Shape.T -> when (r) {
                0 -> listOf(Cell(1, 0), Cell(0, 1), Cell(1, 1), Cell(2, 1))
                1 -> listOf(Cell(1, 0), Cell(1, 1), Cell(2, 1), Cell(1, 2))
                2 -> listOf(Cell(0, 1), Cell(1, 1), Cell(2, 1), Cell(1, 2))
                else -> listOf(Cell(1, 0), Cell(0, 1), Cell(1, 1), Cell(1, 2))
            }
            Shape.J -> when (r) {
                0 -> listOf(Cell(0, 0), Cell(0, 1), Cell(1, 1), Cell(2, 1))
                1 -> listOf(Cell(1, 0), Cell(2, 0), Cell(1, 1), Cell(1, 2))
                2 -> listOf(Cell(0, 1), Cell(1, 1), Cell(2, 1), Cell(2, 2))
                else -> listOf(Cell(1, 0), Cell(1, 1), Cell(0, 2), Cell(1, 2))
            }
            Shape.L -> when (r) {
                0 -> listOf(Cell(2, 0), Cell(0, 1), Cell(1, 1), Cell(2, 1))
                1 -> listOf(Cell(1, 0), Cell(1, 1), Cell(1, 2), Cell(2, 2))
                2 -> listOf(Cell(0, 1), Cell(1, 1), Cell(2, 1), Cell(0, 2))
                else -> listOf(Cell(0, 0), Cell(1, 0), Cell(1, 1), Cell(1, 2))
            }
        }
    }

    data class Piece(
        val shape: Shape,
        val rotation: Int = 0,
        /** Левый верхний угол рамки 4×4 в координатах поля. */
        val x: Int = 3,
        val y: Int = -1,
    ) {
        /** Занятые клетки в координатах поля. */
        val occupied: List<Cell> get() = cells(shape, rotation).map { Cell(x + it.x, y + it.y) }
    }

    data class State(
        /** Уложенные клетки: [ROWS] × [COLUMNS], `null` — пусто. */
        val board: List<List<Shape?>>,
        val piece: Piece?,
        val next: Shape,
        val lines: Int,
        val isOver: Boolean,
        /** Свой генератор, чтобы состояние оставалось воспроизводимым в тестах. */
        val seed: Long,
    ) {
        /** Заработанные бонусы (до дневного лимита — его считает BonusViewModel). */
        val bonuses: Int get() = lines * BONUS_PER_LINE
    }

    fun start(seed: Long = -0x61c8864680b583ebL): State {
        var s = seed
        val first = randomShape(s).also { s = nextSeed(s) }
        val next = randomShape(s).also { s = nextSeed(s) }
        val empty = List(ROWS) { List<Shape?>(COLUMNS) { null } }
        return State(empty, Piece(first), next, lines = 0, isOver = false, seed = s)
    }

    /**
     * Линейный конгруэнтный генератор: нужен воспроизводимый, а не «настоящий»
     * рандом — иначе тест на конкретную раскладку не написать.
     */
    private fun nextSeed(seed: Long): Long = seed * 6364136223846793005L + 1442695040888963407L
    private fun randomShape(seed: Long): Shape {
        val v = ((seed ushr 33) % Shape.entries.size).toInt()
        return Shape.entries[if (v < 0) v + Shape.entries.size else v]
    }

    /** Помещается ли фигура: в поле по бокам/снизу и не наезжает на уложенное. */
    fun fits(piece: Piece, board: List<List<Shape?>>): Boolean {
        for (cell in piece.occupied) {
            if (cell.x < 0 || cell.x >= COLUMNS || cell.y >= ROWS) return false
            if (cell.y < 0) continue            // над полем — нормально при спавне
            if (board[cell.y][cell.x] != null) return false
        }
        return true
    }

    fun move(state: State, dx: Int): State {
        val p = state.piece ?: return state
        if (state.isOver) return state
        val moved = p.copy(x = p.x + dx)
        return if (fits(moved, state.board)) state.copy(piece = moved) else state
    }

    fun rotate(state: State): State {
        val p = state.piece ?: return state
        if (state.isOver) return state
        val turned = p.copy(rotation = (p.rotation + 1) % 4)
        // Простой «отскок» от стен: пробуем сдвинуть на клетку внутрь.
        for (dx in listOf(0, -1, 1, -2, 2)) {
            val candidate = turned.copy(x = turned.x + dx)
            if (fits(candidate, state.board)) return state.copy(piece = candidate)
        }
        return state
    }

    /** Шаг вниз. Если некуда — фигура фиксируется, линии сгорают, приходит новая. */
    fun tick(state: State): State {
        val p = state.piece ?: return state
        if (state.isOver) return state
        val dropped = p.copy(y = p.y + 1)
        return if (fits(dropped, state.board)) state.copy(piece = dropped) else lock(state, p)
    }

    fun hardDrop(state: State): State {
        var p = state.piece ?: return state
        if (state.isOver) return state
        while (true) {
            val down = p.copy(y = p.y + 1)
            if (fits(down, state.board)) p = down else break
        }
        return lock(state, p)
    }

    private fun lock(state: State, piece: Piece): State {
        val board = state.board.map { it.toMutableList() }.toMutableList()
        for (cell in piece.occupied) if (cell.y >= 0) board[cell.y][cell.x] = piece.shape

        // Сгоревшие линии убираем, сверху досыпаем пустые.
        val kept = board.filter { row -> row.any { it == null } }
        val cleared = ROWS - kept.size
        val result = List(cleared) { List<Shape?>(COLUMNS) { null } } + kept.map { it.toList() }

        val seed = nextSeed(state.seed)
        val spawned = Piece(state.next)
        val over = !fits(spawned, result)
        return State(
            board = result,
            piece = if (over) null else spawned,
            next = randomShape(seed),
            lines = state.lines + cleared,
            isOver = over,
            seed = seed,
        )
    }
}
